### PR TITLE
Fix allocation too large buffer for device in sub group test

### PR DESCRIPTION
**Pull request by vladimirkhashev**
The test executed kernel function on a big number of work groups using almost every work items available for work group on particular device. Test should check results for all work items in execution range, a buffer is used to keep test results. Added checking of available buffer size on device. Now if required buffer size exceeds available buffer size, work groups size is reduced.

### CODE REVIEW

File: tests/sub_group/sub_group_api.cpp [hunk 1]
```diff
@@ -170,6 +217,18 @@ TEST_CASE("sub-group api", "[sub_group]") {
...
+    // exceed maximum memory allocation size on device reduce initial
+    // local_size_dim (i.e. current local_size_dim + 1) to nearest multiple two
+    // and if possible, again reduce size by one to attempt to create incomplete
+    // sub-groups
+    size_t required_res_buffer_size =
+        result_buf_size(make_global_range(local_range, local_size_dim).size());
+    if (result_buffer_cant_be_alloc_in_device(required_res_buffer_size)) {
+      local_size_dim = reduce_work_group_size_to_suitable_multiple_two(
+          local_range, local_size_dim + 1);
+      local_size_dim = local_size_dim > 1 ? local_size_dim - 1 : local_size_dim;
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:230]: Unclear. Probably less obfuscated:
```suggestion
      local_size_dim = std::max(local_size_dim - 1, 1);
```

File: tests/sub_group/sub_group_api.cpp [hunk 2]
```diff
@@ -170,6 +217,18 @@ TEST_CASE("sub-group api", "[sub_group]") {
   // if possible, reduce size by one to attempt to create incomplete sub-groups
   if (local_size_dim > 1) {
     --local_size_dim;
+    // if required size for result buffer for current work group dimension size
+    // exceed maximum memory allocation size on device reduce initial
+    // local_size_dim (i.e. current local_size_dim + 1) to nearest multiple two
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:222]:
```suggestion
    // If required size for result buffer for current work group dimension size
    // exceeds maximum memory allocation size on device, reduce initial
    // local_size_dim (i.e. current local_size_dim + 1) to nearest multiple of two
```
But it is unclear what we get at the end. A multiple of a power-of-2?

File: tests/sub_group/sub_group_api.cpp [hunk 3]
```diff
@@ -96,6 +96,53 @@ static size_t get_work_group_size(const sycl::device& device) {
...
+inline bool result_buffer_cant_be_alloc_in_device(size_t required_buffer_size) {
+  return required_buffer_size > max_device_buf_size();
+};
+
+/** Returns global range for given number of work group and work group size. */
+inline sycl::range<3> make_global_range(const sycl::range<3>& work_group_num,
+                                        size_t work_group_size_dim) {
+  return work_group_num * sycl::range<3>{work_group_size_dim,
+                                         work_group_size_dim,
+                                         work_group_size_dim};
```
[the remark this answers is not in the document]
Comment (vladimirkhashev) [reply tests/sub_group/sub_group_api.cpp]: Made different sizes for work group dimensions.

File: tests/sub_group/sub_group_api.cpp [hunk 4]
```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
...
+    size_t dim_size_exp = 0;
+    while (tmp >>= 1) dim_size_exp++;
+    return dim_size_exp;
+  };
+
+  std::transform(desired_work_group_sizes.begin(),
+                 desired_work_group_sizes.end(),
+                 desired_work_group_sizes_exp.begin(), calc_ilog2);
+
+  auto tot_work_items = [](std::vector<size_t> sizes) {
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:93]:
```suggestion
  auto tot_work_items = [](const std::vector<size_t>& sizes) {
```
or even
```suggestion
  auto tot_work_items = [](const auto& sizes) {
```
since you are not checking the number of elements anyway.
But looking at the code, you could move this lambda in the following lambda or just inline it there since it is only used once... :-)

File: tests/sub_group/sub_group_api.cpp [hunk 5]
```diff
@@ -54,10 +54,10 @@ static size_t get_idx(size_t id, member_function member_func) {
 }
 
 /**
- * Returns the largest possible size that each dimension of a three dimensional
+ * Returns the largest possible sizes that each dimension of a three dimensional
  * work-group may have such that it is a power of two and
  * smaller or equal to 1024. */
-static size_t get_work_group_size(const sycl::device& device) {
+static std::vector<size_t> get_work_group_size(const sycl::device& device) {
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:60]: If this function is only for 3D this should appear somehow in the name.

File: tests/sub_group/sub_group_api.cpp [hunk 6]
```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
...
+
+  std::transform(desired_work_group_sizes.begin(),
+                 desired_work_group_sizes.end(),
+                 desired_work_group_sizes_exp.begin(), calc_ilog2);
+
+  auto tot_work_items = [](std::vector<size_t> sizes) {
+    return sizes[0] * sizes[1] * sizes[2];
+  };
+  std::vector<size_t> actual_work_group_sizes{
+      1, 1, static_cast<size_t>(1) << desired_work_group_sizes_exp[2]};
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:97]:
```suggestion
      1, 1, size_t { 1 } << desired_work_group_sizes_exp[2]};
```

File: tests/sub_group/sub_group_api.cpp [hunk 7]
```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
...
+  std::vector<size_t> desired_work_group_sizes = {
+      std::min<size_t>(max_work_item_sizes[0], 1024),
+      std::min<size_t>(max_work_item_sizes[1], 1024),
+      std::min<size_t>(max_work_item_sizes[2], 1024)};
+
+  // calc ilog2(desired_work_group_sizes), desired_work_group_sizes is at least
+  // 1
+  std::vector<size_t> desired_work_group_sizes_exp(3);
+  auto calc_ilog2 = [](size_t dim_size) {
+    size_t tmp = dim_size;
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:83]: I am surprised that you do not directly use `dim_size` instead, since you try to squish a lot the code elsewhere... :-)

File: tests/sub_group/sub_group_api.cpp [hunk 8]
```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
...
-      break;
+  std::vector<size_t> desired_work_group_sizes = {
+      std::min<size_t>(max_work_item_sizes[0], 1024),
+      std::min<size_t>(max_work_item_sizes[1], 1024),
+      std::min<size_t>(max_work_item_sizes[2], 1024)};
+
+  // calc ilog2(desired_work_group_sizes), desired_work_group_sizes is at least
+  // 1
+  std::vector<size_t> desired_work_group_sizes_exp(3);
+  auto calc_ilog2 = [](size_t dim_size) {
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:82]:
```suggestion
  auto ilog2 = [](size_t dim_size) {
```

File: tests/sub_group/sub_group_api.cpp [hunk 9]
```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
...
+    size_t tmp = dim_size;
+    size_t dim_size_exp = 0;
+    while (tmp >>= 1) dim_size_exp++;
+    return dim_size_exp;
+  };
+
+  std::transform(desired_work_group_sizes.begin(),
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:89]: All this PR is hard to follow.
At the end I am not convinced that you really need to rewrite this `desired_work_group_sizes` with a value difficult to follow.
Instead you could just use below `ilog2(desired_work_group_sizes_exp[2])` when you need it and so on, as it looks like they are used only once.

File: tests/sub_group/sub_group_api.cpp [hunk 10]
```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
...
+                 desired_work_group_sizes.end(),
+                 desired_work_group_sizes_exp.begin(), calc_ilog2);
+
+  auto tot_work_items = [](std::vector<size_t> sizes) {
+    return sizes[0] * sizes[1] * sizes[2];
+  };
+  std::vector<size_t> actual_work_group_sizes{
+      1, 1, static_cast<size_t>(1) << desired_work_group_sizes_exp[2]};
+
+  auto increase_dim_size = [&tot_work_items, max_work_group_size](
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:99]: Comment.

File: tests/sub_group/sub_group_api.cpp [hunk 11]
```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
...
+                               std::vector<size_t>& wg_sizes, size_t dim_idx,
+                               size_t dim_exp) {
+    for (size_t curr_exp = 1; curr_exp <= dim_exp; curr_exp++) {
+      wg_sizes[dim_idx] <<= 1;
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:103]: Comment, comment... :-)

File: tests/sub_group/sub_group_api.cpp [hunk 12]
```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
...
+      if (tot_work_items(wg_sizes) > max_work_group_size) {
+        wg_sizes[dim_idx] >>= 1;
+        break;
+      }
     }
-  }
+  };
+
+  increase_dim_size(actual_work_group_sizes, 1,
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:111]: Comment

File: tests/sub_group/sub_group_api.cpp [hunk 13]
```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
...
+                    desired_work_group_sizes_exp[1]);
+  increase_dim_size(actual_work_group_sizes, 0,
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:113]: Comment

File: tests/sub_group/sub_group_api.cpp [hunk 14]
```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
...
+                    desired_work_group_sizes_exp[0]);
+
+  return actual_work_group_sizes;
+}
 
-  return size;
+/** Returns maximum size of buffer keeping test results that can be allocated
+ * on device. */
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:120]:
```suggestion
    on device. */
```

File: tests/sub_group/sub_group_api.cpp [hunk 15]
```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
...
+
+/**
+ * Reduces current size of each dimension of a three dimensional work-group to
+ * nearest suitable power-of-two value to ensure test result buffer will be able
+ * to be allocated on device. Returns resulting sizes for each dimension. */
+std::vector<size_t> reduce_work_group_size_to_suitable_power_of_two(
+    const sycl::range<3>& work_group_num,
+    std::vector<size_t> current_work_group_size_dims) {
+  uint64_t max_dev_buf_size = max_device_buf_size();
+  uint64_t required_res_buffer_size = result_buf_size(
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:158]: I am lost with some of your naming.
I am unsure that `required` is correct. Perhaps `requested` instead?

File: tests/sub_group/sub_group_api.cpp [hunk 16]
```diff
@@ -68,34 +68,106 @@ static size_t get_work_group_size(const sycl::device& device) {
...
+      make_global_range(work_group_num, current_work_group_size_dims).size());
+  uint32_t dim_idx = 0;
+  while (required_res_buffer_size > max_dev_buf_size) {
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:161]: Explain with comments what you are doing here.

File: tests/sub_group/sub_group_api.cpp [hunk 17]
```diff
@@ -166,12 +238,31 @@ TEST_CASE("sub-group api", "[sub_group]") {
...
   // if possible, reduce size by one to attempt to create incomplete sub-groups
-  if (local_size_dim > 1) {
-    --local_size_dim;
+  std::vector<size_t> reduced_local_size_dims{local_size_dims};
+  auto reduce_if_possible = [](size_t& dim_size) {
+    dim_size = std::max(dim_size - 1, size_t(1));
+  };
+  std::for_each(reduced_local_size_dims.begin(), reduced_local_size_dims.end(),
+                reduce_if_possible);
+  // If required size for result buffer for current work group dimension size
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:249]:
```suggestion
  // If requested size for result buffer for current work group dimension size
```

**Review by keryell — COMMENTED** [review]
I have a brain overflow on this PR. :-)
At least the description you made for this PR should appear somewhere in a top-level comment in the code.

File: tests/sub_group/sub_group_api.cpp [hunk 18]
```diff
@@ -155,15 +163,19 @@ std::vector<size_t> reduce_work_group_size_to_suitable_power_of_two(
...
-    size_t& dim_size = current_work_group_size_dims[dim_idx % 3];
-    dim_size = std::max(dim_size >> 1, size_t(1));
-    required_res_buffer_size = result_buf_size(
+  uint32_t iter = 0;
+  // Halve size of each dimension of the work group until result buffer fits
+  // into the available memory on the device, dimension index changes at each
+  // iteration and takes the value 0, 1 or 2 depending on the iteration number
+  // as the remainder of dividing the iteration number by 3
+  while (requested_res_buffer_size > max_dev_buf_size) {
+    size_t& dim_size = current_work_group_size_dims[iter % 3];
```
Comment (keryell) [inline tests/sub_group/sub_group_api.cpp:179]: Oh, I see, it was a reference here!

**Review by keryell — APPROVED** [review]
Great!
This is quite more palatable and enjoyable with the comments. :-)
